src/extract/index.ts

```diff
@@ -0,0 +1,52 @@
+import { zodToJsonSchema } from 'openai-zod-to-json-schema';
+import { type z } from 'zod';
+
+import { type Model, type Msg, MsgUtil } from '../model/index.js';
+
+/**
+ * Extract data using OpenAI structured outputs.
+ *
+ * Always returns an object satisfying the provided Zod schema.
+ */
+export function createExtractFunction<Schema extends z.ZodObject<any>>(args: {
+  /** The ChatModel used to make API calls. */
+  chatModel: Model.Chat.Model;
+  /** A descriptive name for the object to extract. */
+  name: string;
+  /** The Zod schema for the data to extract. */
+  schema: Schema;
+  /** Add a system message to the beginning of the messages array. */
+  systemMessage: string;
+}): (input: string | Msg) => Promise<z.infer<Schema>> {
+  const { chatModel, schema, systemMessage } = args;
+
+  async function runExtract(input: string | Msg): Promise<z.infer<Schema>> {
+    const inputVal = typeof input === 'string' ? input : (input.content ?? '');
+    const messages: Msg[] = [
+      MsgUtil.system(systemMessage),
+      MsgUtil.user(inputVal),
+    ];
+
+    const { message } = await chatModel.run({
```

Comment:
glad the new extract function uses json output by default.

this is one of those areas where supporting non-openai llms isn't going to "just work". maybe we also add an optional `responseFormat?: `json' | 'json_schema' = 'json_schema'` param?

Reply:
the old version of dexter uses function calling for extract functionality, which I never found as reliable as json output (or not json_schema output), so 100% on this change

Reply:
supporting other llms is not a priority atm.